products/cms-integration/debugging-field-guide.md

```diff
@@ -0,0 +1,144 @@
+        1.  [Solution](#orgb85397c)
+    3.  [Transformer trouble](#org4eae7ce)
+    4.  [Handling circular references](#orge68be39)
+        1.  [Solution](#org924393b)
+
+
+
+<a id="orgabae1eb"></a>
+
+# Debugging Field Guide
+
+Oh no. Something broke, didn&rsquo;t it? Here, take this guide. I hope it helps.
+
+
+<a id="org10ab85f"></a>
+
+## JSON schema validation errors
+
+Content is validated against two separate JSON schemas. Once before
+transformation, and once after. When a validation fails, an error will be thrown
+and a message like the following will be logged to the console:
+
+    node.2bddb1a7-6fb1-4503-838d-9c2fcb51c46a (node-health_care_region_page) is invalid after transformation:
+    {
+      "keyword": "type",
+      "dataPath": ".entity.fieldLinkFacilityEmergList",
+      "schemaPath": "#/properties/entity/properties/fieldLinkFacilityEmergList/type",
+      "params": {
+        "type": "object,null"
+      },
+      "message": "should be object,null"
+    }
+    Data found at .entity.fieldLinkFacilityEmergList: []
+
+From this message we can see the following:
+
+-   The exact entity that failed validation
+    -   `node.2bddb1a7-6fb1-4503-838d-9c2fcb51c46a`
+-   Which content model it was
+    -   `node-health_care_region_page`
+-   Which schema failed
+    -   &ldquo;invalid after transformation&rdquo; means the schema in question can be found in
+        `schemas/transformed/node-health_care_region_page.js`
+    -   &ldquo;invalid before transformation&rdquo; would mean the schema could be found in
+        `schemas/raw/node-health_care_region_page.js`
+-   The field that failed validation
+    -   Look in the `dataPath`
+-   The schema rule that was violated
+    -   `schemapath` tells us what rule it was, `params` tells us what the valid
+        options were
+        -   Technically, it tells us what the parameters were for that particular rule
+    -   In this case, we can see it was the `fieldLinkFacilityEmergList`&rsquo;s `type`
+        which was supposed to be either an object or null
+-   The exact error message returned from [AJV](https://github.com/epoberezkin/ajv)
+    -   Look in `message`
+    -   This is the easy way to figure out what the error was, but it&rsquo;s sometimes
+        still mystifying
+-   What the actual data was that failed the validation
+    -   In this case, we can see `.entity.fieldLinkFacilityEmergList` was an empty
+        array
+    -   The schema expected it to be either an object or null, so this failure makes
+        sense
+
+
+<a id="orgf97a4c8"></a>
+
+### Solution
+
+Armed with this information, we can crack open the transformer and see what&rsquo;s
+going on.
+
+**Note:** If a validation error occurs while running the unit tests, you may have
+to scroll up quite a ways to see the actual validation error output. Just look
+for the yellow / gold writing.
+
+
+<a id="org830b381"></a>
+
+## Object doesn&rsquo;t deep equal object??
```

Comment:
This section and the **Solution** one below it are confusing to me.  Are you saying that we may push some changes up without running the unit tests and it will fail in Jenkins?  My workflow for this never really relied on Jenkins - I would (if I remembered) run both of these locally:

- `npm run test:unit src/site/stages/build/process-cms-exports/tests/`
- `npm run build:content:test`

Should we add those to a sort of checklist for adding new content models?

Reply:
Adding a checklist is good, but we can't rely on them being run locally. I specifically remember one instance where it was failing in Jenkins but not on my local branch, which confused me to no end. (That's actually why I added that logging I mentioned in the first place.)

I don't remember why that happened, though.